packages/headless/src/controllers/product-listing/facet/headless-product-listing-facet.ts

```diff
@@ -167,6 +187,77 @@ export function buildFacet(
   };
 }
 
+export interface Facets extends Controller {
+  state: FacetsState;
+}
+
+export interface FacetsState {
+  facets: AutomaticFacet[];
+}
+
+// TODO: We might need to extract a "core" automatic facets controller
+export function buildFacetsV2(engine: ProductListingV2Engine): Facets {
+  if (!loadAutomaticFacetBuilderReducers(engine)) {
+    throw loadReducerError;
+  }
+  const controller = buildController(engine);
+
+  return {
+    ...controller,
+
+    get state() {
+      const facets =
+        engine.state.productListing.facets.map((facet) =>
```

Comment:
I find it really weird to have dynamically generated facet controllers created by `buildAutomaticFacet` inside this controller's state. How could that possibly work? Also, why isn't the AutomaticFacet type not related to Facet?

Should we take a different approach, like having a facet manager that dynamically creates facet controllers by looking at the listings API response?

Reply:
This follows how generated automatic facets work!
> How could that possibly work?

A client would use the facets controller to get a list of facet controllers, and would map this list to UI facet components. Maybe I'm not understanding your question properly?

> Also, why isn't the AutomaticFacet type not related to Facet?

AutomaticFacet is used in the context of generated automatic facets, Facets is used in the context of commerce facets for PLP v2. Once we have a commerce sub-package, this separation should become clearer.

> Should we take a different approach, like having a facet manager that dynamically creates facet controllers by looking at the listings API response?

This approach you propose is the one that is currently used: the state from which we build facet controllers is populated from the product listing response.